### PR TITLE
public key on attachment

### PR DESCRIPTION
**Pull request by Gorash**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by odony** [comment]
I'd like to do a review with you when I'm back in the office.
Some quick early thoughts (still needs a deeper review!):
- `public_key` is a very unfortunate naming choice, because this is not a *public key*.`key` or `token` or something like that would seem to be better
- that field should be protected from read access, shouldn't it?
- wouldn't it be nice to have routes without ID when the key/token is passed? It should be sufficient to retrieve the record (if necessary, we could have a unique constraint to prevent collisions)
- the auto-magical addition of the key within emails looks like it could be exploited, in addition to having a really ugly regex

.. to be continued..

File: odoo/addons/base/ir/ir_attachment.py [hunk 1]
```diff
@@ -436,6 +442,7 @@ def create(self, values):
             values.pop(field, False)
         values = self._check_contents(values)
         self.browse().check('write', values=values)
+        values['access_token'] = str(uuid.uuid4())
```
Comment (Gorash) [inline odoo/addons/base/ir/ir_attachment.py]: remove and add generateToken + js

File: addons/web_editor/static/src/js/editor/rte.summernote.js [hunk 2]
```diff
@@ -1074,8 +1074,14 @@ var SummernoteManager = Class.extend(mixins.EventDispatcherMixin, {
...
         }
         data.__alreadyDone = true;
+
         var mediaDialog = new weWidgets.MediaDialog(this,
-            data.options || {},
+            _.extend({
+                domain: ['|',
+                    '&', ['res_model', '=', data.$editable.data('oe-model')], ['res_id', '=', data.$editable.data('oe-id')],
+                    ['res_model', '=', 'ir.ui.view']],
+                needAccessToken: (data.$editable.data('needAccessToken') === false ? false : true),
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/rte.summernote.js]: `!== false`

File: addons/web_editor/static/src/js/editor/rte.summernote.js [hunk 3]
```diff
@@ -1074,8 +1074,14 @@ var SummernoteManager = Class.extend(mixins.EventDispatcherMixin, {
             return;
         }
         data.__alreadyDone = true;
+
         var mediaDialog = new weWidgets.MediaDialog(this,
-            data.options || {},
+            _.extend({
+                domain: ['|',
+                    '&', ['res_model', '=', data.$editable.data('oe-model')], ['res_id', '=', data.$editable.data('oe-id')],
+                    ['res_model', '=', 'ir.ui.view']],
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/rte.summernote.js]: Why is this needed ?

Comment (Gorash) [reply addons/web_editor/static/src/js/editor/rte.summernote.js]: To have access token on the website and load image from attachments

File: addons/web_editor/static/src/js/editor/rte.summernote.js (hunk 2 above)
Comment (Gorash) [reply addons/web_editor/static/src/js/editor/rte.summernote.js]: of course ^^

File: addons/web_editor/static/src/js/editor/rte.summernote.js (hunk 3 above)
Comment (Gorash) [reply addons/web_editor/static/src/js/editor/rte.summernote.js]: Now, you can have image on the product displayed on the website, and the widget display the attachments on the website and we should not put them on the website anymore. What is easier for users to navigate and share documents, data sheets...

**Comment by Gorash** [comment]
@odony revewed? It's good for you?

**Comment by Gorash** [comment]
@ged-odoo Can you review before merging

**Comment by odony** [comment]
@Gorash I'm reviewing + testing this final version, as I had only looked at an early version...

@tde-banana-odoo I changed base version to 11.0, as this is a complement to 2257583 and meant for 11.0.

**Comment by tde-banana-odoo** [comment]
@odony Sure. I saw a 500+ commits diff on the initial PR and thought it was targetting master.

**Comment by tde-banana-odoo** [comment]
@Gorash while you are at it I don't see any test for the whole "extract base 64 images" feature you added in mail at https://github.com/odoo/odoo/commit/666b59484201004e2cd8890a9c09b6dcf023a725 . As a cleaning of the mail module is pending, I think you should provide a test for your feature. Otherwise I cannot guarantee you won't have to write it again.

Sorry I did not spot it when you did it. Next time you add some JS-like hack in python in mail please ask for a review or at least write a test.

Many thanks,

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 4]
```diff
@@ -181,34 +182,56 @@ var ImageDialog = Widget.extend({
...
-            if (this.media.tagName !== "A") {
-                $('.note-control-selection').hide();
-                media = document.createElement('a');
-                $(this.media).replaceWith(media);
-                this.media = media;
-            }
-            this.media.setAttribute('href', '/web/content/' + img.id + '?unique=' + img.checksum + '&download=true');
-            $(this.media).addClass('o_image').attr('title', img.name).attr('data-mimetype', img.mimetype);
+        var def = $.when();
+        var needAccessToken = img.res_model !== 'ir.ui.view' && this.options.needAccessToken;
```
[the remark this answers is not in the document]
Comment (odony) [reply addons/web_editor/static/src/js/widgets/widgets.js]: Could we add a test somewhere for this feature? So far I have been unable to see it work ;-)

File: odoo/addons/base/ir/ir_attachment.py [hunk 5]
```diff
@@ -438,6 +439,14 @@ def create(self, values):
         self.browse().check('write', values=values)
         return super(IrAttachment, self).create(values)
 
+    @api.one
+    def generate_access_token(self):
+        if self.access_token:
+            return self.access_token
```
Comment (odony) [inline odoo/addons/base/ir/ir_attachment.py]: Seeing this, I think we should restrict the visibility of the `access_token` field to employees only, at the model level. There should be no need for portal/public users to view/modify the token. Either they received it or they didn't.

File: addons/mail/models/mail_message.py [hunk 6]
```diff
@@ -746,7 +746,7 @@ def base64_to_boundary(match):
                         'res_model': 'mail.message',
                     })
                     values['attachment_ids'].append((4, attachment.id))
-                    data_to_url[key] = '/web/image/%s' % attachment.id
+                    data_to_url[key] = '/web/image/%s?access_token=%s' % (attachment.id, attachment.access_token)
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/mail/models/mail_message.py]: oups, I forgot to call generate_access_token to create the attachment from base64

File: addons/web_editor/static/src/js/backend/fields.js [hunk 7]
```diff
@@ -130,6 +130,9 @@ var FieldTextHtmlSimple = basic_fields.DebouncedField.extend(TranslatableFieldMi
         this.$textarea.summernote(this._getSummernoteConfig());
         this.$content = this.$('.note-editable:first');
         this.$content.html(this._textToHtml(this.value));
+        this.$content.data('oe-id', this.recordData.res_id || this.res_id);
+        this.$content.data('oe-model', this.recordData.model || this.model);
+        this.$content.data('needAccessToken', this.model.indexOf('mail.') === 0);
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/backend/fields.js]: `this.model.indexOf('mail.') === 0` => activate for all mail module, but I can move this into mail

File: addons/web_editor/static/src/js/widgets/widgets.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/widgets/widgets.js]: We can't simulate a upload file :(

File: odoo/addons/base/ir/ir_attachment.py (hunk 5 above)
Comment (Gorash) [reply odoo/addons/base/ir/ir_attachment.py]: I add the group 'base.group_user' on the access_token field

**Comment by Gorash** [comment]
@tde-banana-odoo Ok for you ?

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 8]
```diff
@@ -120,7 +120,13 @@ var ImageDialog = Widget.extend({
         this._super.apply(this, arguments);
         this.options = options || {};
         this.accept = this.options.accept || this.options.document ? "*/*" : "image/*";
-        this.domain = this.options.domain || ['|', ['mimetype', '=', false], ['mimetype', this.options.document ? 'not in' : 'in', ['image/gif', 'image/jpe', 'image/jpeg', 'image/jpg', 'image/gif', 'image/png']]];
```
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/widgets/widgets.js]: Mmh ok, you put it in `fetch_existing`... but shouldn't it be here anyway ?

File: addons/web_editor/static/src/js/backend/fields.js [hunk 9]
```diff
@@ -133,6 +134,8 @@ var FieldTextHtmlSimple = basic_fields.DebouncedField.extend(TranslatableFieldMi
         this.$textarea.summernote(this._getSummernoteConfig());
         this.$content = this.$('.note-editable:first');
         this.$content.html(this._textToHtml(this.value));
+        this.$content.data('oe-id', this.recordData.res_id || this.res_id);
+        this.$content.data('oe-model', this.recordData.model || this.model);
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/backend/fields.js]: https://github.com/odoo/odoo/pull/18904/files#diff-60b5ed1959b17b8f763abf3630fe9416L208

File: addons/web_editor/static/src/js/inline.js [hunk 10]
```diff
@@ -53,6 +53,7 @@ snippet_editor.Class.include({
         this._super.apply(this, arguments);
 
         var $editable = $('#editable_area');
+        transcoder.linkImgToAttachmentThumbnail($editable);
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/inline.js]: an error, must use linkImgToAttachmentThumbnail ^^

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 11]
```diff
@@ -139,6 +145,13 @@ var ImageDialog = Widget.extend({
             o.url = this.media.getAttribute('href').replace(/[?].*/, '');
             o.id = +o.url.match(/\/web\/content\/([0-9]*)/, '')[1];
         }
+        var $form = this.$('form[action="/web_editor/attachment/add"]');
+        if (this.options.res_id) {
+            $form.append('<input type="hidden" name="res_id" value="'+this.options.res_id+'"/>');
+        }
+        if (this.options.res_model) {
+            $form.append('<input type="hidden" name="res_model" value="'+this.options.res_model+'"/>');
+        }
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/widgets/widgets.js]: in stable version ?

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 12]
```diff
@@ -181,34 +195,55 @@ var ImageDialog = Widget.extend({
...
-        } else {
-            if (this.media.tagName !== "A") {
-                $('.note-control-selection').hide();
-                media = document.createElement('a');
-                $(this.media).replaceWith(media);
-                this.media = media;
-            }
-            this.media.setAttribute('href', '/web/content/' + img.id + '?unique=' + img.checksum + '&download=true');
-            $(this.media).addClass('o_image').attr('title', img.name).attr('data-mimetype', img.mimetype);
+        var def = $.when();
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/widgets/widgets.js]: if !img.access_token else I create var def; and add an else

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 13]
```diff
@@ -1001,18 +1036,21 @@ var MediaDialog = Dialog.extend({
         return this._super.apply(this, arguments);
     },
     save: function () {
+        var self = this;
+        var _super = this._super.bind(this, arguments);
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/web_editor/static/src/js/widgets/widgets.js]: O.o what?

File: addons/web_editor/controllers/main.py [hunk 14]
```diff
@@ -155,9 +155,11 @@ def attach(self, func, upload=None, url=None, disable_optimization=None, **kwarg
                 'type': 'url',
                 'url': url,
                 'public': True,
-                'res_model': 'ir.ui.view',
+                'res_id': kwargs.get('res_model', 'ir.ui.view') != 'ir.ui.view' and kwargs.get('res_id') or None,
+                'res_model': kwargs.get('res_model', 'ir.ui.view'),
```
Comment (Gorash) [inline addons/web_editor/controllers/main.py]: nb

**Pull request merged**